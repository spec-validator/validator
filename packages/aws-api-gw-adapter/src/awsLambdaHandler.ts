import querystring from 'querystring'

import createHandler, {
  getServerConfigs, ServerConfig, WildCardRequest, WildCardResponse,
} from '@spec-validator/rest-api-server/handler'

export type AwsReq = {
  headers: Record<string, string>,
  path: string,
  queryStringParameters: Record<string, string>,
  httpMethod: string,
  body?: string,
  requestContext: {
    stage: string
  }
}

const getUrl = (awsReq: AwsReq): string =>
  `${awsReq.path}?${querystring.encode(awsReq.queryStringParameters)}`

const toWildCardRequest = (awsReq: AwsReq): WildCardRequest => ({
  body: awsReq.body || '',
  url: getUrl(awsReq),
  method: awsReq.httpMethod,
  headers: awsReq.headers,
})

export const configureBaseUrl = (event: AwsReq): void => {
  const host = event.headers.Host || 'unknown'
  const proto = event.headers['X-Forwarded-Proto'] || 'http'
  const stage = event.requestContext?.stage || 'unknown'

  process.env.REST_API_BASE_URL = host.endsWith('amazonaws.com') ?
    `${proto}://${host}/${stage}` : `${proto}://${host}`
}

export type LambdaHandler<Config extends Partial<ServerConfig>> = ((event: AwsReq) => Promise<WildCardResponse>) & {
  config: Partial<Config>
}

export const createAwsLambdaHandler = <Config extends Partial<ServerConfig>>(config: Config):
  LambdaHandler<Config> => {
  const handle = createHandler(getServerConfigs(config))
  const result = async (event: AwsReq): Promise<WildCardResponse> => {
    configureBaseUrl(event)
    return await handle(toWildCardRequest(event))
  }
  result.config = config
  return result
}
