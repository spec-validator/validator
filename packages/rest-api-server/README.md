# @spec-validator/rest-api-server

The library provides a set of types that specifies HTTP protocol.

Aside of the types it also implements the core logic of an HTTP
server that on the one hand routinely handles incoming requests
and does the routing in a similar fashion `express` does it
and on the other hand leverages type specifications to validate
the incoming requests.

## Routes

The core building block of a server is a route.

The route is a combination of essentially 3 things:

- request object specification
- response object specification
- handler that gets a valid request object as its only parameter and
  returns a valid response object

```ts
import { route, Route } from '@spec-validator/rest-api-server'
import { expectType } from '@spec-validator/test-utils/expectType'
import {
  segmentField as $, choiceField, numberField, stringField,
  unionField, constantField,
} from '@spec-validator/validator/fields'

const routeObject = route({
  request: {
    method: constantField('GET'),
    headers: {
      headerKey: stringField(),
    },
    pathParams: $._('pathKey', stringField()),
  },
  response: unionField(
    {
      statusCode: constantField(201),
      body: numberField(),
    }, {
      statusCode: constantField(202),
      body: choiceField('one', 'two'),
      headers: {
        headerKey: stringField(),
    },
  }),
}).handler(async (request) => ({
  statusCode: 202,
  body: 'one',
  headers: {
    headerKey: request.pathParams.pathKey,
  },
}))

expectType<Route, typeof routeObject>(true)
```

Note, the routes do not handle trafic in any meaningful way,
they are just a specification for the handlers.

### Fluent API for the most common case

There is a fluent API aiming to simplify route configuration in the
most common case (one 200ish response per handler, known status
codes for various request methods).

```ts
import { createRouteCollection  } from '@spec-validator/rest-api-server'

const itemSpec = {
  title: stringField(),
  description: stringField(),
}

// 'routes' is a list that effectively will aggregate the routes
// defined via a '_' helper object
const _ = createRouteCollection()

// As an alternative you may import '_' singleton object that is
// not linked to any routes collection. In that case you would need
// to add routes to the collections explicitly via 'routes.push'

/**
import { _ } from '@spec-validator/rest-api-server'

const routes: Route[] = []

routes.push(_.GET(...).spec(...).handler(...))
*/

const getRoute = _.GET($._('/items')).spec({
  response: {
    body: [itemSpec],
  },
}).handler(async () => ({
  body: [
    {
      title: 'Item N',
      description: 'Description',
    },
  ],
}))

expectType<Route, typeof getRoute>(true)

const postRoute = _.POST($._('/items')).spec({
  request: {
    body: itemSpec,
  },
  response: {
    body: numberField(),
    headers: {
      title: stringField(),
    },
  },
}).handler(async () => ({
  body: 42,
  headers: {
    title: 'Foo',
  },
}))

expectType<Route, typeof postRoute>(true)
```

## Server object

For any arbitrary set of routes it is possible to join those with
extra parameters to craft server configs:

```ts
import { createServer } from '@spec-validator/rest-api-server'

import http from 'http'

const server = createServer({
  baseUrl: 'http://localhost:8080',
  routes: _.routes
})

expectType<http.Server & { serve: () => http.Server }, typeof server>(true)
```

## Running the app

To actually run the server and handle trafic using it,
run `serve` method:

```ts #ignore

server.serve()

```

Once the app is started you may run the following curl command to obtain the
payload from the endpoint:

```
curl -X GET "http://localhost:8080/items"
```

Or post a new entry:

```
curl -X POST "http://localhost:8080/items" -H  "Content-Type: application/json" \
  -d "{\"title\":\"string\",\"description\":\"string\"}"
```

The following call will produce a 400 (validation) error:

```
curl -X POST "http://localhost:8080/items" -H  "Content-Type: application/json" \
  -d "{\"title\":true,\"description\":\"string\"}"
```

## Cheat sheet

Response with cookies:

```ts
import { headerObjectField } from '@spec-validator/rest-api-server/fields'

const withCookies = _.GET($._('/with-cookies')).spec({
  response: {
    body: [itemSpec],
    headers: {
      cookies: headerObjectField({
        cookieStr: stringField(),
        cookieNum: numberField()
      })
    }
  },
}).handler(async () => ({
  body: [
    {
      title: 'Item N',
      description: 'Description',
    },
  ],
  headers: {
    cookies: {
      cookieStr: 'foo',
      cookieNum: 42
    }
  }
}))

expectType<Route, typeof withCookies>(true)
```

Request with authorization:

```ts
const withAuthorization = _.POST($._('/with-cookies')).spec({
  request: {
    headers: {
      authorization: $._('type', stringField())._(' ')._('credentials', stringField())
    }
  }
}).handler(async () => undefined)


expectType<Route, typeof withAuthorization>(true)
```
