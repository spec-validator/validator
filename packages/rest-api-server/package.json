{
  "name": "@spec-validator/rest-api-server",
  "version": "0.0.1",
  "dependencies": {
    "@spec-validator/utils": "0.0.1",
    "@spec-validator/validator": "0.0.1",
    "qs": "^6.9.4"
  },
  "devDependencies": {
    "@spec-validator/test-utils": "0.0.1",
    "@types/express": "^4.17.9",
    "@types/qs": "^6.9.5",
    "@types/supertest": "^2.0.10",
    "express": "^4.17.1",
    "supertest": "^6.0.1"
  }
}
