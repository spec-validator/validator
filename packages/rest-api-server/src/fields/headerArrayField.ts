import { declareField, withErrorDecoration } from '@spec-validator/validator/core'
import { FieldWithRegExpSupport } from '@spec-validator/validator/fields/segmentField'

export interface HeaderArrayField<T> extends FieldWithRegExpSupport<T[]> {
  readonly itemField: FieldWithRegExpSupport<T>,
  readonly regex: RegExp,
  serialize(input: T[]): string
}

// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Language
const SEPARATOR = ', '

export default declareField('@spec-validator/rest-api/fields/headerArrayField', <T>(
  itemField: FieldWithRegExpSupport<T>
): HeaderArrayField<T> => {

  const fieldWithStringSupport = itemField.getStringField()
  const regex = fieldWithStringSupport.regex.source

  const result = {
    itemField,
    regex: new RegExp(`${regex}(${SEPARATOR}${regex})+`),
    validate: (value: any): T[] => {
      if (typeof value !== 'string') {
        throw 'Not a string'
      }
      return value.split(SEPARATOR).map(
        (it, index) => withErrorDecoration(index, () => fieldWithStringSupport.validate(decodeURI(it)))
      )
    },
    serialize: (deserialized: T[]): string =>
      deserialized.map(
        (it, index) => withErrorDecoration(index, () => encodeURI(fieldWithStringSupport.serialize(it)))
      ).join(SEPARATOR),
  } as HeaderArrayField<T>

  result.getStringField = () => result

  return result
})
