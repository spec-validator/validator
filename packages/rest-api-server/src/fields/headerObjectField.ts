import { Any } from '@spec-validator/utils/util-types'
import { declareField, TypeHint, withErrorDecoration } from '@spec-validator/validator/core'
import {
  FieldWithStringInputSupport,
} from '@spec-validator/validator/fields/segmentField'

export type HeaderSpec<DeserializedType extends Record<string, Any> = Record<string, Any>> = {
  [P in keyof DeserializedType]: FieldWithStringInputSupport<DeserializedType[P]>
}

export interface HeaderObjectField<Spec extends HeaderSpec = HeaderSpec>
  extends FieldWithStringInputSupport<TypeHint<Spec>> {
  readonly objectSpec: Spec,
  readonly separator: string
  readonly regex: RegExp
}

// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cookie
export default declareField('@spec-validator/rest-api/fields/headerObjectField', <Spec extends HeaderSpec>(
  objectSpec: Spec,
  separator='; '
): HeaderObjectField<Spec> => {

  const regex = '.+=.+'

  const result = {
    objectSpec,
    separator,
    regex: new RegExp(`${regex}(${separator}${regex})*`),
    validate: (value: any): TypeHint<Spec> => {
      if (typeof value !== 'string') {
        throw 'Not a string'
      }
      const payload = Object.fromEntries(value.split(separator)
        .map(it =>it.split(/=(.*)/, 2)))

      return Object.fromEntries(Object.entries(objectSpec).map(([key, valueSpec]) => [
        key, withErrorDecoration(key, () => valueSpec.getFieldWithRegExp().validate(payload[key])),
      ])) as TypeHint<Spec>
    },
    serialize: (deserialized: TypeHint<Spec>): string =>
      Object.entries(objectSpec)
        .map(([key, valueSpec]) => [
          key, withErrorDecoration(key, () => valueSpec.getFieldWithRegExp().serialize(deserialized[key])),
        ])
        .map(([key, value]) => `${key}=${value}`)
        .join(separator),
  } as HeaderObjectField<Spec>

  result.getFieldWithRegExp = () => result

  return result
})
