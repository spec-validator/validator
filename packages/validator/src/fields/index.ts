import arrayField from './arrayField'
import booleanField from './booleanField'
import numberField from './numberField'
import objectField from './objectField'
import stringField from './stringField'
import choiceField from './choiceField'
import withDefault from './withDefault'
import optional from './optional'
import unionField from './unionField'
import constantField from './constantField'
import $ from './segmentField'
import wildcardObjectField from './wildcardObjectField'
import undefinedField from './undefinedField'
import dateField from './dateField'

export {
  dateField,
  arrayField,
  booleanField,
  numberField,
  objectField,
  stringField,
  choiceField,
  withDefault,
  optional,
  unionField,
  constantField,
  $,
  wildcardObjectField,
  undefinedField,
}
