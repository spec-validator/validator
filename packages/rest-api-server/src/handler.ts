import querystring from 'querystring'

import { validate, serialize, TypeHint } from '@spec-validator/validator'
import { RequestSpec, ResponseSpec, route as declareRoute, Route, StringMapping, VoidOrUndefined } from './route'
import { cached } from '@spec-validator/utils/utils'
import { HtmlSerialization, JsonSerialization, SerializationFormat } from './serialization'
import { Any, WithoutOptional } from '@spec-validator/utils/util-types'
import { SegmentField, ConstantField, constantField } from '@spec-validator/validator/fields'
import { getFieldForSpec } from '@spec-validator/validator/interface'
import { Field } from '@spec-validator/validator/core'

export type ServerConfig = {
  readonly baseUrl: string,
  readonly serializationFormats: [SerializationFormat, ...SerializationFormat[]],
  readonly encoding: BufferEncoding,
  readonly appErrorStatusCode: number,
  readonly reportError: (error: unknown) => Promise<void>
  readonly routes: Route[],
}

const splitPath = (url: string) => {
  const [pathParams, queryParams] = url.split('?', 2)
  return {
    pathParams,
    queryParams: querystring.decode(queryParams),
  }
}

const DEFAULT_SERVER_CONFIG: ServerConfig = {
  baseUrl: 'http://localhost:8000',
  serializationFormats: [new JsonSerialization(), new HtmlSerialization()],
  encoding: 'utf-8',
  appErrorStatusCode: 500,
  reportError: (error: unknown) => {
    console.error(error)
    return Promise.resolve(undefined)
  },
  routes: [],
}

export const getServerConfigs = <C extends Partial<ServerConfig>> (
  serverConfig: C
): ServerConfig => ({
    ...DEFAULT_SERVER_CONFIG,
    ...serverConfig,
  })

const matchRoute = (
  request: WildCardRequest,
  requestSpec: Pick<RequestSpec, 'method' | 'pathParams'>,
): boolean => {
  try {
    validate(
      {
        method: requestSpec.method,
        pathParams: requestSpec.pathParams,
      },
      {
        method: request.method,
        pathParams: splitPath(request.url).pathParams,
      }
    )
    return true
  } catch (_) {
    return false
  }
}

const withAppErrorStatusCode = async <T>(
  statusCode: number,
  inner: () => Promise<T> | T
): Promise<T> => {
  try {
    return await inner()
  } catch (error) {
    if (error.statusCode && error.isPublic) {
      throw error
    } else {
      throw {
        statusCode,
        isPublic: statusCode < 500,
        reason: error,
      }
    }
  }
}

export type WildCardRequest = {
  body: string,
  headers: Record<string, any>,
  method: string,
  url: string
}

export type WildCardResponse = {
  body?: string,
  headers?: Record<string, any>,
  statusCode: number
}

const getSerializationMapping = (
  serializationFormats: SerializationFormat[],
): Record<string, SerializationFormat> =>
  // To prevent serialization mapping construction on every request
  cached('formats', () => Object.fromEntries(serializationFormats.map(it =>
    [it.mediaType,  it]
  )))

export const firstHeader = (value: string | string[] | undefined): string | undefined => {
  if (!value) {
    return undefined
  } if (typeof value === 'string') {
    return value
  } else {
    return value[0]
  }
}

const ANY_MEDIA_TYPE = '*/*'

const getMediaType = (
  config: ServerConfig,
  requestIn: WildCardRequest,
  headerKey: string,
  fallback?: string
) => {
  const serializationFormats = getSerializationMapping(config.serializationFormats)
  const types = firstHeader(requestIn.headers[headerKey]) || config.serializationFormats[0].mediaType
  const ttype = types
    .split(',')
    .map(it => it.split(';')[0])
    .find((it) => it === ANY_MEDIA_TYPE || serializationFormats[it]) || fallback

  const eventualType = ttype === ANY_MEDIA_TYPE ? config.serializationFormats[0] : ttype && serializationFormats[ttype]
  if (!eventualType) {
    throw {
      statusCode: 415,
      isPublic: true,
      reason: `Not supported '${headerKey}': ${types}`,
    }
  }
  return eventualType
}

// eslint-disable-next-line max-statements
const handleRoute = async (
  config: ServerConfig,
  route: {
    request: {
      headers?: Field<TypeHint<Route['request']['headers']>>,
      queryParams?: Field<TypeHint<Route['request']['queryParams']>>,
      method: Route['request']['method'],
      pathParams: Route['request']['pathParams'],
      body?: Route['request']['body']
    },
    response: Route['response'],
    handler: Route['handler']
  },
  wildcardRequest: WildCardRequest
): Promise<WildCardResponse> => {

  wildcardRequest.headers['accept'] = wildcardRequest.headers['accept'] || wildcardRequest.headers['content-type']

  const _getMediaType = getMediaType.bind(null, config, wildcardRequest)

  const contentType = _getMediaType('content-type')

  const request = await withAppErrorStatusCode(
    400,
    async () => validate(route.request, {
      ...wildcardRequest,
      ...splitPath(wildcardRequest.url),
      body: route.request.body ?
        await withAppErrorStatusCode(400, () => contentType.deserialize(wildcardRequest.body)) :
        undefined,
    }, true),
  )

  // This cast is totally reasoanble because in the interface we exclude
  // null values.
  const handler = route.handler as unknown as (req: TypeHint<RequestSpec>) => Promise<TypeHint<ResponseSpec>>

  // TODO: no any here
  const resp = serialize(route.response, await withAppErrorStatusCode(
    config.appErrorStatusCode,
    async () => await handler(request as any)
  ) as any) as any

  const accept = _getMediaType('accept')

  return {
    body: accept.serialize(resp.body),
    headers: {
      ...resp.headers,
      'content-type': accept.mediaType,
    },
    statusCode: resp.statusCode,
  }
}

export type Placeholder = Any

type RequestSpecMethod = Omit<RequestSpec, 'method' | 'pathParams'>

type ResponseSpecMethod = Omit<ResponseSpec, 'statusCode'>

// Make it fluid API - to make things work with autocomplete
export const withMethod = <
  Method extends string,
  OkStatusCode extends number
> (method: Method, okStatusCode: OkStatusCode, routes?: Route[]) =>
  <PathParams extends StringMapping | undefined = StringMapping | undefined> (pathParams: SegmentField<PathParams>): {
    spec:  <
      ReqSpec extends RequestSpecMethod = RequestSpecMethod,
      RespSpec extends ResponseSpecMethod = ResponseSpecMethod,
      > (spec: {
      request?: ReqSpec,
      response?: RespSpec
    }) => ({
      handler: (
        handler: (request: WithoutOptional<TypeHint<
          ReqSpec
          & {
              readonly method: ConstantField<Method>,
              readonly pathParams: SegmentField<PathParams>
            }
        >>
      ) => Promise<
        VoidOrUndefined<WithoutOptional<TypeHint<RespSpec>>>
      >) => Route
    })
  } => ({
      spec: (spec) => ({ handler: ( handler ) => {
        const entry = declareRoute({
          request: {
            ...(spec.request || {}),
            method: constantField(method),
            pathParams: pathParams as SegmentField<StringMapping | undefined>,
          },
          response: {
            ...(spec.response || {}),
            statusCode: constantField(okStatusCode),
          },
        }).handler(
          async (req) => ({ ...((await handler(req as any)) as any), statusCode: okStatusCode })
        )
        routes?.push(entry)
        return entry
      },
      }),
    })

// We don't want to define the types explicitly because those definitions are quite large
// - better to infer them
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const createRouteCollection = (routes: Route[] | undefined = []) => [({
  GET: withMethod('GET', 200, routes),
  HEAD: withMethod('HEAD', 200, routes),
  POST: withMethod('POST', 201, routes),
  PUT: withMethod('PUT', 204, routes),
  DELETE: withMethod('DELETE', 204, routes),
  PATCH: withMethod('PATCH', 204, routes),
}), routes] as const

export const _ = createRouteCollection(undefined)[0]

const jsonSerialization = new JsonSerialization()

// eslint-disable-next-line max-statements
export default (
  config: ServerConfig
): ((request: WildCardRequest) => Promise<WildCardResponse>) => {

  const routes = config.routes.map(it => ({
    ...it,
    path: {
      method: it.request.method,
      pathParams: it.request.pathParams,
    },
    request: {
      ...it.request,
      headers: getFieldForSpec(it.request.headers, true),
      queryParams: getFieldForSpec(it.request.queryParams, true),
    },
  }))

  // eslint-disable-next-line max-statements
  return async (request: WildCardRequest): Promise<WildCardResponse> => {

    try {
      const route = routes.find(it => matchRoute(request, it.path))

      if (!route) {
        return {
          statusCode: 404,
        }
      }

      return await handleRoute(config, route, request)
    } catch (error) {
      const statusCode = error.statusCode || config.appErrorStatusCode
      if (error.isPublic) {
        const accept = getMediaType(config, request, 'accept', jsonSerialization.mediaType)
        return {
          statusCode,
          body: accept.serialize(error),
        }
      } else {
        await config.reportError(error)
        return {
          statusCode,
        }
      }
    }

  }
}
