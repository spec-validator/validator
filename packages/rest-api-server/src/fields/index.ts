import headerArrayField from './headerArrayField'
import headerObjectField from './headerObjectField'

export { headerObjectField, headerArrayField }
