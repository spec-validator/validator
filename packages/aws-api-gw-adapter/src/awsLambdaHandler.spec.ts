import { _ } from '@spec-validator/rest-api-server/handler'
import { segmentField as $, numberField, stringField } from '@spec-validator/validator/fields'
import { createAwsLambdaHandler, configureBaseUrl } from './awsLambdaHandler'

const awsRequest = {
  headers: {
    keyIn: 'valueIn',
    'content-type': 'application/json',
    Host: 'example.com',
  },
  path: '/items/42',
  queryStringParameters: { qkey: 'qvalue' },
  httpMethod: 'POST',
  requestContext: {
    stage: 'dev',
  },
  body: JSON.stringify({
    title: 'Title',
    description: 'Description',
  }),
}


describe('configureBaseUrl', () => {
  beforeEach(() => {
    delete process.env.REST_API_BASE_URL
  })

  afterEach(() => {
    delete process.env.REST_API_BASE_URL
  })

  it('fills empty placeholders with strings', () => {
    configureBaseUrl({
      ...awsRequest,
      headers: {},
    })
    expect(process.env.REST_API_BASE_URL).toEqual(
      'http://unknown'
    )
  })

  it('uses staged aws url in case of amazonaws.com urls', () => {
    configureBaseUrl({
      ...awsRequest,
      headers: {
        Host: 'api-gw.amazonaws.com',
      },
    })
    expect(process.env.REST_API_BASE_URL).toEqual(
      'http://api-gw.amazonaws.com/dev'
    )
  })

  it('uses defined domain if any', () => {
    configureBaseUrl({
      ...awsRequest,
      headers: {
        Host: 'example.com',
        'X-Forwarded-Proto': 'https',
      },
    })
    expect(process.env.REST_API_BASE_URL).toEqual(
      'https://example.com'
    )
  })
})

test('entrypoint', async () => {
  expect(await createAwsLambdaHandler({
    routes: [
      _.POST($._('/items/')._('id', numberField())).spec(
        {
          request: {
            queryParams: {
              qkey: stringField(),
            },
            body: {
              title: stringField(),
              description: stringField(),
            },
            headers: {
              keyIn: stringField(),
            },
          },
          response: {
            body: {
              id: numberField(),
              title: stringField(),
              key: stringField(),
            },
          },
        },
      ).handler(
        async (req) => ({
          body: {
            id: req.pathParams.id,
            title: req.body.title,
            key: req.queryParams.qkey,
          },
        })
      ),
    ],
  })(awsRequest)).toMatchSnapshot()

})
