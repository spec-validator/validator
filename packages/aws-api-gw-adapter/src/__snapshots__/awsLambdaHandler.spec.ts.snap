// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`entrypoint 1`] = `
Object {
  "routes": Array [
    Object {
      "handler": [Function],
      "request": Object {
        "body": Object {
          "description": Object {
            "getStringField": [Function],
            "regex": /\\.\\*/,
            "serialize": [Function],
            "type": "@spec-validator/validator/fields/stringField",
            "validate": [Function],
          },
          "title": Object {
            "getStringField": [Function],
            "regex": /\\.\\*/,
            "serialize": [Function],
            "type": "@spec-validator/validator/fields/stringField",
            "validate": [Function],
          },
        },
        "headers": Object {
          "keyIn": Object {
            "getStringField": [Function],
            "regex": /\\.\\*/,
            "serialize": [Function],
            "type": "@spec-validator/validator/fields/stringField",
            "validate": [Function],
          },
        },
        "method": Object {
          "constant": "POST",
          "getStringField": [Function],
          "regex": /POST/,
          "serialize": [Function],
          "type": "@spec-validator/validator/fields/constantField",
          "validate": [Function],
        },
        "pathParams": SegmentField {
          "field": Object {
            "canBeFloat": false,
            "regex": /\\\\d\\+/,
            "serialize": [Function],
            "signed": false,
            "type": "@spec-validator/validator/fields/numberField",
            "validate": [Function],
          },
          "key": "id",
          "parent": SegmentField {
            "field": undefined,
            "key": "/items/",
            "parent": SegmentField {
              "field": undefined,
              "key": "",
              "parent": undefined,
              "type": "@spec-validator/validator/fields/segmentField",
            },
            "type": "@spec-validator/validator/fields/segmentField",
          },
          "type": "@spec-validator/validator/fields/segmentField",
        },
        "queryParams": Object {
          "qkey": Object {
            "getStringField": [Function],
            "regex": /\\.\\*/,
            "serialize": [Function],
            "type": "@spec-validator/validator/fields/stringField",
            "validate": [Function],
          },
        },
      },
      "response": Object {
        "body": Object {
          "id": Object {
            "canBeFloat": false,
            "getStringField": [Function],
            "regex": /\\\\d\\+/,
            "serialize": [Function],
            "signed": false,
            "type": "@spec-validator/validator/fields/numberField",
            "validate": [Function],
          },
          "key": Object {
            "getStringField": [Function],
            "regex": /\\.\\*/,
            "serialize": [Function],
            "type": "@spec-validator/validator/fields/stringField",
            "validate": [Function],
          },
          "title": Object {
            "getStringField": [Function],
            "regex": /\\.\\*/,
            "serialize": [Function],
            "type": "@spec-validator/validator/fields/stringField",
            "validate": [Function],
          },
        },
        "statusCode": Object {
          "constant": 201,
          "getStringField": [Function],
          "regex": /201/,
          "serialize": [Function],
          "type": "@spec-validator/validator/fields/constantField",
          "validate": [Function],
        },
      },
    },
  ],
}
`;

exports[`entrypoint 2`] = `
Object {
  "body": "{\\"id\\":42,\\"title\\":\\"Title\\",\\"key\\":\\"qvalue\\"}",
  "headers": Object {
    "content-type": "application/json",
  },
  "statusCode": 201,
}
`;
